extensions/ql-vscode/src/databaseFetcher.ts

```diff
@@ -0,0 +1,114 @@
+import * as fetch from "node-fetch";
+import * as unzipper from "unzipper";
+import { ExtensionContext, Uri, ProgressOptions, ProgressLocation, commands, window } from "vscode";
+import * as fs from "fs-extra";
+import * as path from "path";
+import { DatabaseManager } from "./databases";
+import { ProgressCallback, showAndLogErrorMessage, withProgress } from "./helpers";
+
+
+
+export default async function runCommand(dbm: DatabaseManager, ctx: ExtensionContext) {
+  try {
+    const databaseUrl = await window.showInputBox({
+      prompt: 'Enter URL of zipfile of database to download'
+    });
+
+    validateUrl(databaseUrl);
+
+    if (databaseUrl) {
+      const progressOptions: ProgressOptions = {
+        location: ProgressLocation.Notification,
+        title: 'Adding database from URL',
+        cancellable: false,
+      };
+      await withProgress(progressOptions, async progress => await databaseFetcher(databaseUrl, dbm, ctx, progress));
+      commands.executeCommand('codeQLDatabases.focus');
+    }
+  } catch (e) {
+    showAndLogErrorMessage(e.message);
+  }
+}
+
+async function databaseFetcher(
+  databaseUrl: string,
+  databasesManager: DatabaseManager,
+  ctx: ExtensionContext,
+  progressCallback: ProgressCallback
+): Promise<void> {
+  progressCallback({
+    maxStep: 3,
+    message: 'Downloading database',
+    step: 1
+  });
+  const storagePath = ctx.storagePath || ctx.globalStoragePath;
```

Comment:
maybe factor out this `this.ctx.storagePath || this.ctx.globalStoragePath`, and the one that occurs in `databaseFetcher.ts`, into a function in `databaseFetcher.ts` that takes an extension context and returns that, just to reduce the probability that those ever get out of sync in the future.